src/main/java/net/snowflake/ingest/streaming/internal/BdecParquetReader.java

```diff
@@ -0,0 +1,213 @@
+package net.snowflake.ingest.streaming.internal;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.parquet.hadoop.ParquetReader;
+import org.apache.parquet.hadoop.api.InitContext;
+import org.apache.parquet.hadoop.api.ReadSupport;
+import org.apache.parquet.io.DelegatingSeekableInputStream;
+import org.apache.parquet.io.InputFile;
+import org.apache.parquet.io.SeekableInputStream;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.io.api.Converter;
+import org.apache.parquet.io.api.GroupConverter;
+import org.apache.parquet.io.api.PrimitiveConverter;
+import org.apache.parquet.io.api.RecordMaterializer;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.MessageType;
+
+public class BdecParquetReader implements AutoCloseable {
+  private final ParquetReader<List<Object>> reader;
+
+  public BdecParquetReader(byte[] data) throws IOException {
+    reader =
+        new ParquetReader.Builder<List<Object>>(new BdecInputFile(data)) {
+          @Override
+          protected ReadSupport<List<Object>> getReadSupport() {
+            return new BdecReadSupport();
+          }
+        }.build();
+  }
+
+  public List<Object> read() throws IOException {
+    return reader.read();
+  }
+
+  @Override
+  public void close() throws IOException {
+    reader.close();
+  }
+
+  /**
+   * Reads the input data using Parquet reader and writes them using a Parquet Writer.
+   *
+   * @param data input data to be read and written into outputWriter
+   * @param outputWriter output parquet writer
+   */
+  public static void readFileIntoWriter(byte[] data, BdecParquetWriter outputWriter) {
```

Comment:
The multichannel case per table may degrade now because there is no better known way to merge channels for Parquet. The reason is that the channels cannot be easily merged because our XP scanner cannot scan multi-page Parquet files and there is no way to merge multiple pages into one. I suspect generally because of fancy encodings in Parquet.

This was not a problem before because we buffered Java object lists that are easy to merge but very suboptimal to buffer memory-wise. I think it is still better to optimise for one page case as this is probably more frequent case and eventually what we should do anyways when XP supports multi-page file scans. Then there is only a mem copy of the page buffers during the merge, no per-row parse/copy.

See also our previous discussion in [the design doc](https://docs.google.com/document/d/1xYpR7FQhZMMlbXGLzix-pfAtPrMj2m6VvU3zzNvuHq4/edit#heading=h.l5pdprh4q06f).

Anyways, I agree we should have a perf throughput test to see the difference. `MultiChannelsMultiTablesRunner.kts` looks to cover this case, just as discussed with Toby offline, the number of rows should be 1-10 Million. Although, the existing performance degradation in SDK 5 can mask problems with Parquet vs Arrow tests. Hence, I suggest the existing degradation is firstly fixed as Lukas is already investigating it.

Reply:
> Anyways, I agree we should have a perf throughput test to see the difference.

I'll do that as discussed offline. I'll add a parameter in this PR to "protect" the optimization in case we wan to disable it.